Adds Gyro Support to MoveTank

nshenoy opened the pull request:
Adds the following gyro supported functions to MoveTank:
- follow_gyro_angle: angle following function that uses the same PID algorithm as the line follower
 to have the robot drive straight at a specific angle.
- pivot_gyro: pivots the robot to a specific angle
- calibrate_gyro: flips the mode of the sensor to zero-out the gyro to reduce chance of drift

These changes may partially address #620 as it follows the intent of that issue in that it uses the gyro for accurate driving.

nshenoy committed:
Adds Gyro Support to MoveTank

Adds the following gyro supported functions to MoveTank:
- follow_gyro_angle: angle following function that uses the same PID algorithm as the line follower
 to have the robot drive straight
 at a specific angle.
- pivot_gyro: pivots the robot to a specific angle
- calibrate_gyro: flips the mode of the sensor to zero out the gyro to reduce
 chance of drift

nshenoy committed:
Fixes whitespace

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -1857,6 +1866,9 @@ def __init__(self, left_motor_port, right_motor_port, desc=None, motor_class=Lar
         self.cs = None
 
+        # gyro sensor used by follow_gyro_angle()
+        self.gyro = None
```
How does the gyro get initialized?

WasabiFan replied:
I see from the examples it seems the intent is for the user to set it themselves. While this might be fine in general, I worry it will cause problems because a) we don't do this elsewhere AFAIK and b) it probably won't show up well in the documentation. I'm thinking it might be better to have a setter method or a property for it.

nshenoy replied:
You're correct, the intent for the user to set it. I followed the same pattern used for the color sensor `cs` attribute. I can make `gyro` an explicit property, but will that make it inconsistent though?

WasabiFan replied:
Hmmm... That seems odd to me. Relatedly, I'm a bit bugged by the fact that we're putting so many different optional things in the `MoveTank` class. This isn't an issue specific to your PR; I'm just thinking through how we really want to do it.

How about, at least for now, we make it a property (getter/setter pair with `_gyro` private member internally) with a docstring so that it shows up in the documentation, and also change the color sensor to match? I don't think it will break anyone's existing code.

nshenoy replied:
That's a good idea. I'll work on making them both properties in a non-breaking way.

dwalton76 replied:
Setting the self.cs this way is my fault. We should use a setter/getter for that one too

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``target_angle`` is the angle we want drive
```
missing `to`

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+        ``sleep_time`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
```
How does it perform with a 10ms loop time? Does it actually run at that frequency or is it limited by the execution time?

nshenoy replied:
Excellent question. My tests have primarily been to use the default of 10ms. When using the gyro in EV3-G, we've had to introduce some amount of small sleep for the gyro to chillax. This allows users to tweak this if their gyro needs more rest between samples.

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+
+        .. code:: python
+
+            from ev3dev2.motor import OUTPUT_A, OUTPUT_B, MoveTank, SpeedPercent, follow_for_ms
```
where is `follow_for_ms` defined?

nshenoy replied:
That's defined on [line 1826](https://github.com/ev3dev/ev3dev-lang-python/blame/3d22941975ca2b017812198120868f2cd4333e3c/ev3dev2/motor.py#L1826)  introduced by #630 .

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+                    follow_for=follow_for_ms,
+                    ms=4500
+                )
+            except Exception:
```
Probably good to catch the `FollowGyroAngleErrorTooFast` error here explicitly

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+            try:
+                # Follow the line for 4500ms
+                tank.calibrate_gyro()
+                tank.follow_gyro_angle(
```
Where is the target angle?

What's the expected use case for this? Is the idea that you'd first retrieve the current angle then follow that? I'm in particular wondering how you know what target angle to use.

nshenoy replied:
Typo in my example code. Good catch!

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+            left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+            right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            if left_speed > MAX_SPEED:
```
Looking at it... what's the motivation for failing if the speed is too high as opposed to just capping it at that max speed?

I think we did this for the line following too and not I'm not sure why.

nshenoy replied:
It's a good way for the user to know that they're potentially shooting themselves in the foot. Going over a max could result in the sensors and algorithm to not correct themselves within the feedback loop's quantum.

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``target_angle`` is the angle we want drive
```
this is the angle we want to _rotate to_, right?

WasabiFan commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+                    speed=SpeedPercent(5),
+                    target_angle(30)
+                )
+            except Exception:
```
same note as above regarding exception type

nshenoy replied:
I'll remove. This method doesn't throw an exception unless someone forgot to initialize the gyro. But this sample properly initializes it so there will be zero chance of the AssertionError being raised.

WasabiFan reviewed (commented):
Thanks for the contribution! Some various minor comments/questions.

nshenoy committed:
Fixes some PR feedbac

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,178 @@ def follow_line(self,
+
+        self.stop()
+
+    def pivot_gyro(self,
```
How about `turn_to_angle_gryo`? That would be more consistent with `MoveDifferential.turn_to_angle()`

nshenoy replied:
Makes sense.

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2111,178 @@ def follow_line(self,
+
+        while not target_reached:
+            current_angle = self.gyro.angle
+            if (current_angle >= target_angle-2 and current_angle <= target_angle+2):
```
could do `if abs(current_angle - target_angle) <= 2:` to get this down to a single check

nshenoy replied:
Good catch! I'll fix this.

nshenoy committed:
Refactors Color and Gyro sensors to be properties

Added explicit color and gyro setters/getters

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -1854,8 +1863,23 @@ def __init__(self, left_motor_port, right_motor_port, desc=None, motor_class=Lar
+    # color sensor used by follow_line()
+    @property
+    def cs(self):
+        return self.__cs
```
For cs and for gyro is the double leading underscore intentional or should this be the more standard single leading underscore?  I didn't realize double leading underscore had a special meaning but then read this
https://dbader.org/blog/meaning-of-underscores-in-python

I'm not sure we need the name mangling for these...I don't think we use this for any other private variables.

nshenoy replied:
I'd argue that the double underscore is the more pythonic way of keeping the private variable truly private and having accessor methods. But I agree that if this doesn't fit the conventions of the rest of the project, then we should make it consistent with a single underscore. Let me know what you prefer.

WasabiFan replied:
Personally I'd say use a single underscore to be consistent within this codebase. I didn't know about the double-underscore's behavior -- Python always has some new implicit behavior to deal with! _shakes fist_

dwalton76 replied:
I like sticking with single underscore for consistency

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2123,178 @@ def follow_line(self,
+
+        ``follow_for`` is called to determine if we should keep following the
+            desired angle or stop.  This function will be passed ``self`` (the current
+            ``MoveWithGyroTank`` object). Current supported options are:
```
`MoveWithGyroTank` -> `MoveTank` (there is no `MoveWithGyroTank` class)

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2123,178 @@ def follow_line(self,
+            left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+            right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            if left_speed > MAX_SPEED:
```
I think this needs to be `if abs(left_speed)...` in case the motor is being asked to move backwards at an impossible speed.  That seems very unlikely but you never know.

Same for the right_speed check on 2224

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2123,178 @@ def follow_line(self,
+            )
+        """
+        assert self.__gyro, "GyroSensor must be defined"
+
```
We should assert that ``speed`` is <= the max speed of the motors

nshenoy replied:
Good idea! I'll add that.

dwalton76 commented on ev3dev2/motor.py:
```diff
@@ -2099,6 +2123,178 @@ def follow_line(self,
+
+        while not target_reached:
+            current_angle = self.__gyro.angle
+            if (abs(current_angle - target_angle) <= 2):
```
Would pass the 2 into this method as an arg `wiggle_room=2` and then compare against `wiggle_room` here...that way the user can control how accurate they want it to be.

nshenoy committed:
Modifies private vars to use single underscore

Addresses code review feedback including:
- making private members use single instead of double underscore
- `wiggle_room` added to  `turn_to_angle_gyro` for user to control accuracy

dwalton76 commented:
Looks good to me...merging. 
Thanks for the code :)
